### PR TITLE
#1122 Navigating unmounts components

### LINKED ISSUE #1122: Navigating unmounts components
## Describe the bug

When navigating in the app, each page on the `NavigatigationStack` is `unmounted` every time navigation occurs.  This is a significant preformance degradation, as when a user is navigating to a page 3 deep, there are three large unmounting, then remounting process' occuring, including an initial render. These are extremely slow process' compared to a simple re-render.

### To reproduce
1. Put a log in `componentDidUnmount` of some pages you will navigate to
2. Start navigating

1. Use a function which uses `runWithLoadingIndicator`
2. See pages in the stack (a part from mSupplyAppContainer) unmount

I.e - Add a log in `CustomerInvoices` page and navigate to a `CustomerInvoice`, see `CustomerInvoicesPage` unmount.

### Expected behaviour

Components should not be unmounting.

### Version and device info

- App version: I would say all, but possibly 2.3+
- Tablet model:  All
- OS version:  All

### Additional context

**This is about the unmounting of components in the navigation stack rather than the many re-renders of mSupplyMobileApp**

I believe there are two issues here. The large re-rendering is because of `props` being passed. Simple re-rendering of pages should not cause unmounting of components that are already mounted.

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1122 

## Change summary

- Whenever `mSupplyMobileAppContainer` re-renders, `ReduxNavigator` re-renders. When `ReduxNavigator` re-renders, a new `ReduxContainer` is created and generates a new navigation stack.
- When the new navigation stack is created, all components currently in that stack are mounted, and when the old navigation stack gets garbage collected, the components in the stack are unmounted.

Example of navigating from `CustomerInvoicePage` back to `CustomerInvoicesPage`

image

1. Render root component (Creating a new navigation stack, re-renders due to prop changes from being subscribed to redux which passes new navigation state in props)
2. Mount a new `MenuPage` (From the new navigation stack)
3. Render the new `MenuPage`  
4. Mount a new `CustomerInvoicesPage` (From the new navigation stack)
5. Render a new `CustomerInvoicesPage`
6. Unmount the old `MenuPage` (Garbage collection)
7. Unmount the old `CustomerInvoicesPage` (Garbage collection)
8. Unmount the `CustomerInvoicePage` (Garbage collection)
9. Render the new `CustomerInvoicesPage` (New top of the stack)

#### With this change

image

1. Render root component
2. Render `MenuPage`
3. Render `CustomerInvoicesPage`
4. Render `CustomerInvoicePage`
5. Render `CustomerInvoicesPage`
6. Unmount `CustomerInvoicePage`
7. Render root component

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Navigating does not throw any errors
- [ ] Navigating does not unmount components currently in the navigation stack

### Related areas to think about

**Note:** this does not fix the rendering problems described in #1123 - that would require more changes, this issue was solely for the expensive unmounting process. 

This does significantly improve navigation performance, but not as much as fixing the rendering issues would.

### CODE REVIEW

File: src/navigation/ReduxNavigation.js [hunk 1]
```diff
@@ -4,9 +4,11 @@ import { connect } from 'react-redux';
 import { createReduxContainer } from 'react-navigation-redux-helpers';
 import AppNavigator from './Navigator';
 
+const ReduxNavigator = createReduxContainer(AppNavigator, 'root');
+
 const ReduxNavigation = props => {
   const { nav, dispatch, screenProps } = props;
-  const ReduxNavigator = createReduxContainer(AppNavigator, 'root');
+
```
Comment (wlthomson) [inline src/navigation/ReduxNavigation.js]: Can make this a nice one-liner:

```
const ReduxNavigation = ({ nav, dispatch, screenprops }) => <ReduxNavigator state={nav} dispatch={dispatch} screenProps={screenProps} />;
```

**Comment by joshxg** [comment]
I can't replicate this at all - is this simply navigating between the pages or is there another step? It looks as though the update is occurring on a modal rather than the `CustomerInvoicePage` also. There's an issue somewhere to replace the modal component as it is antiquated and has problems with updating it's own state

**Comment by wlthomson** [comment]
> I can't replicate this at all - is this simply navigating between the pages or is there another step? It looks as though the update is occurring on a modal rather than the `CustomerInvoicePage` also. There's an issue somewhere to replace the modal component as it is antiquated and has problems with updating it's own state

Sorry, I forgot to specify that the error case occurs when navigating back from the new invoice modal, I've edited accordingly. I assumed this was still an issue, as the page is getting unmounted while navigating.

**Comment by joshxg** [comment]
Oh right. At the moment modals are not closed when clicking the back button, rather the normal navigation behaviour occurs (which I hope to change soon). So it does make sense that the component unmounts in that case as it is being popped off the navigation stack. I am not getting the unmounted component noop, but I am quite certain that it is because of the modal component that we use. Which manipulates state directly rather than through `setState`, so often gets into an inconsistent state and doesn't know if it's Arthur or Martha.

Wrapping the component with this conditional would fix this problem: 

```
{isCreatingInvoice && (
          <SelectModal
            isOpen={isCreatingInvoice}
            options={database.objects('Customer')}
            placeholderText={modalStrings.start_typing_to_select_customer}
            queryString="name BEGINSWITH[c] $0"
            sortByString="name"
            onSelect={name => {
              this.onNewInvoice(name);
              this.setState({ isCreatingInvoice: false });
            }}
            onClose={() => this.setState({ isCreatingInvoice: false })}
            title={modalStrings.search_for_the_customer}
          />
        )}
```

as we would ensure the mounting and unmounting are controlled by our state rather than the modals internal state. However this page will be refactored shortly, with all the others #1125

**Comment by wlthomson** [comment]
> Oh right. At the moment modals are not closed when clicking the back button, rather the normal navigation behaviour occurs (which I hope to change soon). So it does make sense that the component unmounts in that case as it is being popped off the navigation stack. I am not getting the unmounted component noop, but I am quite certain that it is because of the modal component that we use. Which manipulates state directly rather than through `setState`, so often gets into an inconsistent state and doesn't know if it's Arthur or Martha.
> 
> Wrapping the component with this conditional would fix this problem:
> 
> ```
> {isCreatingInvoice && (
>           <SelectModal
>             isOpen={isCreatingInvoice}
>             options={database.objects('Customer')}
>             placeholderText={modalStrings.start_typing_to_select_customer}
>             queryString="name BEGINSWITH[c] $0"
>             sortByString="name"
>             onSelect={name => {
>               this.onNewInvoice(name);
>               this.setState({ isCreatingInvoice: false });
>             }}
>             onClose={() => this.setState({ isCreatingInvoice: false })}
>             title={modalStrings.search_for_the_customer}
>           />
>         )}
> ```
> 
> as we would ensure the mounting and unmounting are controlled by our state rather than the modals internal state. However this page will be refactored shortly, with all the others #1125

Got it, makes sense. Updated to pass tests.

**Pull request merged**